src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java

```diff
@@ -559,6 +561,12 @@
      * @since 1.4
      */
     private Map<String, String> globalKnownLicenses;
+    
+    boolean isAggregatorBuild;
```

Comment:
Can `isAggregatorBuild` ever be false? Both subclasses set it to `true` or I oversee something? Can perhaps `isAggregatorBuild` be eliminated through encapsulation?

Reply:
It is false when using `AddThirdPartyPlugin` as a goal. It is only true when the plugin is called via `AggregateAddThirdPartyMojo`. The line in `AddThirdPartyPlugin` where it is set to true is only called through `AggregateAddThirdPartyMojo` (it essentially invokes the plugin once for each child project in the reactor).

Regarding whether it can be eliminated, I don't think so without a larger refactoring. The problem is that the `init` method needs to behave differently when used as part of an `AggregateAddThirdPartyMojo` invocation, both when called directly on the `AggregateAddThirdPartyMojo` class, and the `AddThirdPartyMojo` instances it creates. `init` doesn't take parameters, and the interface is defined at the `AbstractLicenseMojo` level. I didn't want to start rewriting these classes to not use inheritance so heavily, so adding a field instead was the only solution I could think of.

Reply:
> It is false when using AddThirdPartyPlugin as a goal.

I see `isAggregatorBuild = true` in `AddThirdPartyPlugin` https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-f71923bddd2327bb3c7b3c62c80878c2R467

> Regarding whether it can be eliminated, I don't think so without a larger refactoring. The problem is that the init method needs to behave differently when used as part of an AggregateAddThirdPartyMojo invocation, both when called directly on the AggregateAddThirdPartyMojo class, and the AddThirdPartyMojo instances it creates. init doesn't take parameters, and the interface is defined at the AbstractLicenseMojo level.

Not sure I understand all details, but can't you define something like `protected abstract LoadedProjectDependencies loadDependencyArtifacts()` in `AbstractAddThirdPartyMojo`, implement it differently in the two subclasses and replace this conditional https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-0bfb2b7e70b9abacb4339555c7b49e81R668 with the follwing? 

```
dependencyArtifacts = loadDependencyArtifacts()
```

Reply:
BTW, if you really intended to have isAggregatorBuild = false in AddThirdPartyPlugin https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-f71923bddd2327bb3c7b3c62c80878c2R467 and all tests are passing, then there is probably a test missing that would fail due to this mistake?

Reply:
> I see isAggregatorBuild = true in AddThirdPartyPlugin https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-f71923bddd2327bb3c7b3c62c80878c2R467

Yes. This line is in the `initFromMojo` method, which is used by `AggregatorAddThirdPartyMojo` to initialize the `AddThirdPartyMojo` instances it creates (here: https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-38b03aa7a6c189ea0f7e814313a14682R215). It isn't used when running `AddThirdPartyMojo` as a goal directly.

This is also why doing something like `protected abstract LoadedProjectDependencies loadDependencyArtifacts()` doesn't work. The behavior of `loadDependencyArtifacts()` would need to do one thing when `AddThirdPartyMojo` runs it as part of the `addThirdParty` goal, while it needs to do something else when `AggregatorAddThirdPartyMojo` instantiates a bunch of `AddThirdPartyMojo`s to run for its child projects.

Trying to give an overview:
* Running the `addThirdParty` goal from Maven: `init` has to act as if `isAggregatorBuild == false`. 
* Running `aggregateAddThirdParty` will run `init` once through `AggregatorAddThirdPartyMojo` for the parent project, where `isAggregatorBuild == true`, and then run `init` once per child project through `AddThirdPartyMojo.initFromMojo` where `isAggregatorBuild == true`.

Reply:
> Yes. This line is in the `initFromMojo` method, which is used by `AggregatorAddThirdPartyMojo` to initialize the `AddThirdPartyMojo` instances it creates (here: https://github.com/mojohaus/license-maven-plugin/pull/148/files#diff-38b03aa7a6c189ea0f7e814313a14682R215). It isn't used when running `AddThirdPartyMojo` as a goal directly.

OK, thanks for the explanation, I was not aware of the fact that `AggregatorAddThirdPartyMojo`  instantiates `AddThirdPartyMojo`s.

Given what you say, `isAggregatorBuild` seems to be significant for `AddThirdPartyMojo` only (because it is always true for `AggregatorAddThirdPartyMojo`). Could we perhaps move `isAggregatorBuild` to `AddThirdPartyMojo` and introduce a `protected abstract LoadedProjectDependencies loadDependencyArtifacts()` in `AbstractAddThirdPartyMojo` and let  it alwas return

* `dependenciesTool.loadProjectArtifacts( localRepository, remoteRepositories, project, reactorProjectDependencies )` from `AggregatorAddThirdPartyMojo` and
* the current conditional from `AddThirdPartyMojo`? 

I think this would improve the readability of the code.

Reply:
Made this change. Turns out the method wasn't necessary in the aggregator mojo, so I removed it from there.